### ISSUE TITLE
Types of a fake partition in map_partition inferring are not exactly correct

### ISSUE THREAD

**Issue by frol** [dask/dask#1512]
Consider the following code:

```python
import dask.dataframe
import pandas

def qq(partition_df):
    print(partition_df)
    print(partition_df['X'].dtype)
    return partition_df

df = dask.dataframe.from_pandas(pandas.DataFrame({'X': range(100)}, dtype='int32'), npartitions=1)
print(df['X'].dtype)
df.map_partitions(qq).compute()
```

The output is:

```
float32
float64
float32
```

while I expect to see:

```
float32
float32
float32
```

Can the types be just copied from the original dataframe? How can I distinguish fake partitions from true ones? I have a heavy computation which I would rather avoid on fake partitions. In Dask 0.10 fake partition was empty, but Dask 0.11 passes a partition with two records in it.

**Comment by mrocklin** [comment]
You would provide a `meta=` keyword argument to the `map_partitions` call.   

@jcrist maybe we should have a docpage about Dask.dataframe metadata?  Questions like these have been coming up a bit recently.  It'd be nice to have something to point people towards.

**Comment by frol** [comment]
@mrocklin I would like to have types and column names in one place, that is inside the map function... With Dask 0.10 I could do the following:

```python
def map_func(df):
    my_columns = ['a', 'b']
    if df.empty:
        return pandas.DataFrame([], columns=my_columns, dtype='float32')
    ...
    new_df = pandas.DataFrame(data, columns=my_columns, dtype='float32')
    ...
    return new_df
```

**Comment by jcrist** [comment]
> @jcrist maybe we should have a docpage about Dask.dataframe metadata? Questions like these have been coming up a bit recently. It'd be nice to have something to point people towards.

That's a good idea, I'll write that up.

> Consider the following code: ...

I'm seeing `int64` where it should be `int32` (fixing now), but I'm not seeing floats anywhere. My experience:

```python
```python
In [1]: %load temp.py

In [2]: # %load temp.py
import dask.dataframe
import pandas

def qq(partition_df):
    print(partition_df['X'].dtype)
    return partition_df

df = dask.dataframe.from_pandas(pandas.DataFrame({'X': range(100)}, dtype='int32'), npartitions=3)
res = df.map_partitions(qq)
   ...:
int64

In [3]: t = res.compute()
int32
int32
int32

In [4]: t.dtypes
Out[4]:
X    int32
dtype: object
```

**Comment by mrocklin** [comment]
> @mrocklin I would like to have types and column names in one place, that is inside the map function... 

I share that desire.  It competes with other desires though.  If you want dask.dataframe to be able to reason about your dtypes then we have three options:

1.  You need to provide them to it with a `meta=` keyword argument
2.  We call your function on a bit of data to guess what the dtypes are
3.  We inspect your function to determine your intent

Obviously 3 is very hard.  Until we figure something else out (suggestions and pull requests very welcome here) we're stuck with 1 or 2

Previous versions of dask.dataframe allowed the creation of dask.dataframes without known dtypes.  This is no longer allowed as it created some difficult to maintain situations.

**Comment by jcrist** [comment]
The bug causing numeric dtype size inference issues is fixed in #1513.

```python
In [1]: %load test
test.h5  test.py

In [1]: %load temp.py

In [2]: # %load temp.py
import dask.dataframe
import pandas

def qq(partition_df):
    print(partition_df['X'].dtype)
    return partition_df

df = dask.dataframe.from_pandas(pandas.DataFrame({'X': range(100)}, dtype='int32'), npartitions=3)
res = df.map_partitions(qq)
   ...:
int32

In [3]: res.dtypes
Out[3]:
X    int32
dtype: object

In [4]: _ = res.compute()
int32
int32
int32
```

Note however that dtype inference in pandas is tricky due to all the implicit casting that pandas will do internally. For most simple user functions we should be able to guess dtype correctly, but there will be cases where we fail, and you should provide the `meta` keyword (see docstring here: http://dask.pydata.org/en/latest/dataframe-api.html#dask.dataframe.DataFrame.map_partitions). I'll write up a section for the docs in a separate PR.

**Comment by frol** [comment]
I have found that I can distinguish fake partitions by looking at `partition_df.is_copy` value, which is `None` for fake partitions and `weakref` for true partitions!
